### DISCUSSION TITLE
Training Paddle Slim for mobile but it is too bad

### DISCUSSION THREAD

**Discussion by hoangtrongbinh1111** [PaddlePaddle/PaddleOCR#14350] [Q&A]
I have trained Paddle slim with 3 millions images for OCR task with Vietnamese text, the accuracy is maybe about 90% with 60 epochs, but when I use paddle_lite_opt to convert to *.nb file and test it in android_demo, Paddle Lite version 2.14rc, it only show 1 character in every images. Please give me instructions to enhace it~ Many thanks
![image](https://github.com/user-attachments/assets/48a4bcbf-50f4-4428-8a60-d6c46c8e02e9)

**Comment by hoangtrongbinh1111** [comment]
![image](https://github.com/user-attachments/assets/31848749-e916-4e8b-b5ab-2332dbb27491)
This is my training process

**Comment by SWHL** [comment]
I think we can first confirm whether the trained model can be inferred normally on the validation set, and then see whether the converted model is correct.

**Reply by hoangtrongbinh1111** [reply]
In the web base it is still ok, but in the mobile after converting, it is too bad :(

**Reply by hoangtrongbinh1111** [reply]
In android demo from Paddle OCR, I only change PaddleLite 2.10 to 2.14, and I used pretrained model from chinese, it is so good. And I retrained with my data, with their configs, and convert to nb file, then apply to android demo, it is so poor. It cannot recog any character.

**Comment by GreatV** [comment]
It seems like your issue is related to the conversion and deployment of your PaddleOCR model using Paddle Lite. Based on the provided details, here are some suggested steps to improve your deployment and resolve the issue of only one character being recognized:

---

### 1. **Verify Your Model Conversion Process**
- When converting your trained PaddleSlim model to a `.nb` file using `paddle_lite_opt`, ensure that:
  - Both the `--model_file` (network structure) and `--param_file` (weights) are correctly specified.
  - Use the `--optimize_out_type=naive_buffer` option to optimize for mobile deployment.
  - Specify the `--valid_targets=arm` since you are targeting an Android device.
  
  Example command for conversion:
  ```bash
  paddle_lite_opt --model_file=./your_model.pdmodel \
                  --param_file=./your_model.pdiparams \
                  --optimize_out=./optimized_model \
                  --valid_targets=arm \
                  --optimize_out_type=naive_buffer
  ```

---

### 2. **Check the Dictionary File**
- OCR models rely on a dictionary file (e.g., `ppocr_keys_v1.txt`) for mapping predictions to characters. If your OCR task involves Vietnamese text, ensure:
  - The dictionary file matches your training setup (i.e., it includes all Vietnamese characters used during training).
  - Replace or modify the dictionary file if needed. You can find dictionary examples in `PaddleOCR/ppocr/utils/`.

---

### 3. **Update the Model's Input Settings**
- Ensure the input image dimensions and preprocessing steps are consistent across training, conversion, and inference. For PaddleOCR:
  - Set the `rec_image_height` to `48` for PP-OCRv3 models or `32` for PP-OCRv2 models in the configuration file (`config.txt`).
  - Example:
    ```txt
    rec_image_height 48
    max_side_len 960
    det_db_thresh 0.3
    det_db_box_thresh 0.5
    det_db_unclip_ratio 1.6
    ```

---

### 4. **Test with Pretrained Models**
- Before testing your custom model, verify the inference pipeline with pretrained models provided by PaddleOCR (e.g., PP-OCRv3 slim models). This will help identify if the issue is specific to your model or the deployment process.
  - Detection Model: [Download here](https://paddleocr.bj.bcebos.com/PP-OCRv3/chinese/ch_PP-OCRv3_det_slim_infer.nb)
  - Recognition Model: [Download here](https://paddleocr.bj.bcebos.com/PP-OCRv3/chinese/ch_PP-OCRv3_rec_slim_infer.nb)

---

### 5. **Debugging the Android Demo**
- Ensure the `.nb` models, test images, dictionary files, and configuration files are correctly placed in the Android demo's `debug` folder.
- Push the necessary files to your Android device using `adb`:
  ```bash
  adb push ./debug /data/local/tmp/
  adb shell
  cd /data/local/tmp/debug
  export LD_LIBRARY_PATH=${PWD}:$LD_LIBRARY_PATH
  ```
- Run the inference demo, specifying the detection and recognition models:
  ```bash
  ./ocr_db_crnn system ch_PP-OCRv3_det_slim_opt.nb ch_PP-OCRv3_rec_slim_opt.nb \
                 ch_ppocr_mobile_v2.0_cls_slim_opt.nb arm8 INT8 10 1 ./test_image.jpg \
                 config.txt ppocr_keys_v1.txt True
  ```

---

### 6. **Validate Your Model's Output**
- If the model still outputs only one character:
  - Double-check the post-processing logic in your PaddleOCR pipeline (e.g., CRNN post-processing for recognition).
  - Ensure the detection model's output is properly passed to the recognition model.
  - Verify that the bounding boxes generated by the detection model are correct.

---

### 7. **Check Paddle Lite Version Compatibility**
- Ensure the Paddle Lite version used (`2.14rc` in your case) is compatible with the `.nb` model generated. If issues persist, try using a stable version like `2.11` or `2.12` for better compatibility.

---

### 8. **Quantization Considerations**
- If you quantized your model during training or conversion, ensure your hardware supports INT8 inference. If not, try using FP32 precision instead:
  ```bash
  ./ocr_db_crnn system ch_PP-OCRv3_det_slim_opt.nb ch_PP-OCRv3_rec_slim_opt.nb \
                 ch_ppocr_mobile_v2.0_cls_slim_opt.nb arm8 FP32 10 1 ./test_image.jpg \
                 config.txt ppocr_keys_v1.txt True
  ```

---

### 9. **Further Resources**
- Refer to the [PaddleOCR Android deployment guide](https://github.com/PaddlePaddle/PaddleOCR/tree/release/2.6/deploy/android_demo) for additional setup details.
- Check the [PaddleOCR Discussions](https://github.com/PaddlePaddle/PaddleOCR/discussions/14095) for similar issues or raise your question there.

---

By following these steps, you should be able to identify and resolve the issue with your PaddleOCR mobile deployment. If the issue persists, consider sharing more details about your training and deployment environment.

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*

**Reply by hoangtrongbinh1111** [reply]
I used version 2.14rc and didn’t encounter any errors. However, when using the android_demo, I only changed the version of PaddleLite from 2.10 to 2.14 in the Android demo and replaced it with the quantized model I trained, it failed to recognize. On the other hand, with the chPPOCRv3 model you provided, there were no issues.

**Comment by hoangtrongbinh1111** [comment]
paddle_lite_opt --model_file=./latest/inference.pdmodel  --param_file=./lite/latest/inference.pdiparams  --optimize_out=./latest --valid_targets=arm  --optimize_out_type=naive_buffer => this is my conversion with 2.14rc. there are no problems
